### ISSUE TITLE
[Request] minetest.override_entity

### ISSUE THREAD

**Issue by Rui-Minetest** [minetest/minetest#2675]
I want to be able to add the function to the registered entity.

・Example

```
minetest.override_entity("aa:bb", {
    mesh = "ccc.b3d",
    on_rightclick = function(self)
        self.anim = aaaa
    end,
})
```

**Comment by est31** [comment]
At the very least, provide a description what the requested function should do. Just a title isn't sufficient.

**Comment by C1ffisme** [comment]
Um, yeah, and speak in correct English sentences. "and should be able to do it"?

**Comment by Rui914** [comment]
↑↑↑
OK?

**Comment by est31** [comment]
I still don't get what you want. Describe what parameters the function has, and what it does.

**Comment by Rui914** [comment]
OK?

**Comment by est31** [comment]
no

**Comment by est31** [comment]
Yes

**Comment by C1ffisme** [comment]
NOW we understand you. Looks useful!!! Maybe someone wants to make a mod that requires zombies from one mod to be a lot stronger. (Maybe for balance, like if the mod being made is an armor mod.)

**Comment by 4aiman** [comment]
Wait... I seem to remember seeing ```minetest.register_entity(":some_mod:some_name", definition)```
And that worked.

Otherwise, I think that it's easier to change the settings of the original mod.
That is, unless you want to override some particular entity, not the class.

**Comment by Rui914** [comment]
@4aiman
When using it, other definitions are reset.

**Comment by 4aiman** [comment]
@Rui914 
>When using it, other definitions are reset.

What do you mean by that?

**Comment by Rui914** [comment]
if
minetest.register_entity("aa:bb", {
    mesh = "ccc.b3d",
    hp_max = 5,
})
and
minetest.register_entity(":aa:bb", {
    hp_max = 5,
})
then
minetest.registered_entities["aa:bb"]
{
    hp_max = 5,
}

I'm sorry to poor English!

**Comment by 4aiman** [comment]
Well, that's a natural thing ;)

That's why I kept telling everyone to copy nodedef before overriding.
Same here - use copy function to copy the initial definition and then change it as you like and override.
Here, have an example of [deepcopy](http://lua-users.org/wiki/CopyTable)

**Comment by C1ffisme** [comment]
Best that we add this function rather than copy. It's not smart programming to copy everything. What if it changes?

Even if you just got all the settings from the origional, you could still add or remove variables, and it's tedious to do that anyway. Do you want to do this?:

Origional Mod:

```
minetest.register_entity("mod:blah",{
mesh = "blah.x",
texture = "blah.png",
max_hp = 40,
... 30 extra entries later ...
})
```

New Mod:

```
local entity  = minetest.registered_entities[blah]
minetest.register_entity(":mod:blah",{
mesh = entity.mesh,
texture = entity.texture,
max_hp = entity.max_hp,
... 30 similar entries later ...
})
```

This is a way to make programmers, REALLY TIRED.

**Comment by 4aiman** [comment]
@C1ffisme 
You've being misguided by someone.

If you want to change all those ```30 similar entries``` then a function won't be able to help you.
how is
```minetest.override_entity(':mod:blah',{hp=entity.hp, texture=entity.texture, max_hp = entity.max+ph, ...})```
better than
```
minetest.register_entity(":mod:blah",{mesh = entity.mesh, texture = entity.texture, max_hp = entity.max_hp, ...})
```
Care to find 1 difference? ;)

You'll have to override all the properties you want to.
If case you want to change all of them, then why override at all?
Make a fork - ppl *will* have to download your mod anyway.
Moreover, your mod *will* be incompatible with the old one. And by "incompatible" I mean that the ones who would like to use your version will have to abandon the original. *Because* of override. 

On the other hand, should you create a fork w/o changing the names of entities - the result will be the same. Only ppl *won't* be in need to download the original mod.


**The bottom line is**: you don't want to use override function and to create a dependent mod if you're trying to change a lot of properties. Because there won't be many differences.
In case you'll need to override just a couple of properties in lots of definitions, then override may come in handy - you won't need to deepcopy stuff.
Just remember, that Lua uses pointers and *not* deepcopying may cost you quite a time to get it why your mod behaves strange.



PS: If you're tired - go to sleep and sleep well. That's my rule of a thumb. Quite handy one ;)

**Comment by Amaz1** [comment]
@4aiman I'm probably missing something, but I think that minetest.override_entry doesn't need all the old defs in.. So, to change the hp of an entity it would simply be
```lua
minetest.override_entity('mod:blah', {
    hp = 10
})
```
instead of 
```lua 
minetest.register_entity(":mod:blah", {
    mesh = entity.mesh, 
    texture = entity.texture, 
    hp = 10, 
...})
```
At least, that's how (I think) it works for nodes!

**Comment by C1ffisme** [comment]
@4aiman You missunderstand me. I don't want to copy all those entries, I want to change one, and keep all the others. However, we can't do that at the moment.

**Comment by 4aiman** [comment]
@Amaz1, the example I gave was based on @C1ffisme's example. 
It looked as if it was a try to change multiple properties.  It was also said by @C1ffisme that it is a "tiresome" way to override 30 or more properties.

I merely demonstrated that for changing 30 or so properties, the ```minetest.register_entity(':mod:blah',def)``` have little differences in comparison to a desired ```minetest.override_entity``` function.

@C1ffisme , if you want to change just one then:
-  deepcopy
- change one property
- override with ```minetest.register_entity(':mod:blah',def)```
That is.

Also, Amaz, how can I get in touch with you? PM me @minetest.net, please.

You don't need additional C++ code. Your function may be added to your mod in Lua. 

So far you can't register or override anything after the game has been launched.
Thus, it doesn't matter how slow deepcopy will be - it'll be run only once per entity that is being overrided *before* the game starts.


Just make it a part of YOUR api:
```
function deepcopy(orig)
    local orig_type = type(orig)
    local copy
    if orig_type == 'table' then
        copy = {}
        for orig_key, orig_value in next, orig, nil do
            copy[deepcopy(orig_key)] = deepcopy(orig_value)
        end
        setmetatable(copy, deepcopy(getmetatable(orig)))
    else -- number, string, boolean, etc
        copy = orig
    end
    return copy
end

function minetest.override_entity(name,properties)
    if not minetest.registered_entities[name] then return end
    local def = deepcopy(minetest.registered_entities[name])
    for k,v in pairs(properties) do
         def[k] = v
    end
    minetest.register_entity(':'..name,def)
end
```

Then you'll be able to use that function like so:
```
minetest.override_entity('adbs:zombie',{max_hp = 20})
minetest.override_entity('adbs:wormhole',{attack_power = 10, max_hp = 3})
```
(^Should anyone want this, the license would be LGPLv2+)
================================================

My point is: you don't have to wait for anyone. 
You can do it on your own right now.
And Minetest... it's up to the devs whether adopt this code or not.
While they're thinking you'll be using the desired function.
(BTW, minetest.swap_node was called default.hacky_swap_node at some point ;)

**Comment by C1ffisme** [comment]
Ok, I see your reasoning, but still, why not have it in normal minetest? Less programming means simpler mods. Why not just put that code in builtin?

**Comment by Rui914** [comment]
Why don't you add it in builtin?

**Comment by 4aiman** [comment]
> Why don't you add it in builtin?

I *totally support* things like mentioned in this feature request to be part of the "standard" package, but...

Me? How can I possibly do that? :)
I'm not a "core dev" (Thank you, The Entity of Entropy, The Entity of Water and All Liquids, The Entity of The Moon and Divine Punishment and you, The Entity of Wind!)


But in all seriousness, this is how *I* see the future:

Get some libraries (or APIs from mods) like ```vector```, add a ```vector = minetest``` to the first line of init.lua and try to make *that compilation* to be popular. 
Minetest seemed to declare to be "open" to include different subgames, I guess a good toolset will have much more chances to get added to the package.

**Comment by Rui914** [comment]
It is very useful.
You really should have pull request.

**Comment by 4aiman** [comment]
There's no need to do that. 
The only thing needed is to copy&paste the code which is already provided.
Who and "where to" would possibly like to paste it - that is not of my concern.

I don't want to force anyone doing anything.
That being said, merging a pull request would take more steps than copying&pasting.
I can see @est31 here.
Why not ask him to merge this WTFPL snippet?

**Comment by Rui914** [comment]
Yes.
By the way, I have a question.
Can I change registered_* tables by mods?
If it can't be, then mods can't use override_entity.

**Comment by 4aiman** [comment]
You can, but only before the game has been started.
I.e. you can do it while your mod is being loaded but not after that.
You also need to call minetest.register_* in order to notify th engine you've changed smth within some definition.

Otherwise you'll get unknown item/node or the change won't apply (it would be read by the Lua, but C++ part won't know anything).

That's the reason there are so many additional tools in every other "enchantments" mod.


But that only concerns things that go to the C++ part.
One can easily change the colour of a sheep in run-time and preserve that with on_activate and get_staticdata.

**Comment by ShadowNinja** [comment]
You can actually do something as simple as `minetest.registered_entities["foo"].hp = 10`, since entity propertied are loaded from the registration table every time they're spawned.  This should probably be abstracted away by a override function though.

**Comment by 4aiman** [comment]
```minetest.registered_entities["foo"].hp = 10```
So, changing in realtime will lead to a bug where entities would use the old definitions until re-loaded.
And one can't create a separate registration in real-time either...
